compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -81,9 +84,9 @@ class ClassfileParser(
   private def currentIsTopLevel(using Context) = classRoot.owner.is(Flags.PackageClass)
 
   private def mismatchError(className: SimpleName) =
-    throw new IOException(s"class file '${in.file.canonicalPath}' has location not matching its contents: contains class $className")
+    throw new IOException(s"class file '${classfile.canonicalPath}' has location not matching its contents: contains class $className")
 
-  def run()(using Context): Option[Embedded] = try {
+  def run()(using Context): Option[Embedded] = try withReader(classfile) { (using in) =>
     report.debuglog("[class] >> " + classRoot.fullName)
     parseHeader()
     this.pool = new ConstantPool
```

Comment:
Something I'm not sure about is whether like in Scala 2 `pool` should be set to `null` after `parseClass`. In Scala 2 the ClassfileParser is a field of the ClassfileLoader so it's important to not keep reference to unused things as that might prevent garbage-collection, in Dotty ClassfileParser is only ever allocated in a method of `ClassfileLoader` and we don't keep any direct reference to it, so it should be garbage-collected as soon as we're done using it, unless we accidentally keep a reference to it in a completer we set, but that should not be necessary and we should make sure that's not the case.

Reply:
Good point. Regarding the `pool`, a more principled way is to pass it around, as we did for `DataReader`. As both have exactly the same life duration, it might be possible to create a facade to combine them --- ConstantPool can be reused too.

The refactoring will make it significantly different from Scala2, I think it's better to leave it to another PR for experiments.

Reply:
OK, so until these refactorings are done it's important to set `pool = null` after `parseClass()` if we want to allow the ConstantPool to be garbage-collected while we keep references to the ClassfileParser.

Reply:
Yes, good point, that is done now.